File: SegmentTrees/segment_tree.py
__author__ = "shashwat tiwari"
__email__  = "shashwat1791@gmail.com"

import math
import sys

class SegmentTree:

	def __init__(self):
		pass

	def get_next_power_of_2(self, number):
		count = 0
		while number != 0:
			number = number >> 1;
			count += 1

		return int(math.pow(count,2))

	def initialize_segment_tree(self, arr):
		st_length = self.get_next_power_of_2(len(arr))*2-1
		segment_tree = [sys.maxint]*st_length

		self.build_segment_tree(segment_tree, arr, 0, len(arr)-1, 0)
		return segment_tree

	def build_segment_tree(self, segment_tree, input, start, end, pos):

		if start == end:
			segment_tree[pos] = input[start]
			return 

		mid = (start+end)//2	

		# recurse to the left and right half
		self.build_segment_tree(segment_tree, input, start, mid, 2*pos+1)
		self.build_segment_tree(segment_tree, input, mid+1, end, 2*pos+2)

		segment_tree[pos] = min(segment_tree[2*pos+1],segment_tree[2*pos+2])
		return 

	def range_min_query(self, segment_tree, start, end, qs, qe):
		return self.range_min_query_helper(segment_tree, qs, qe, start, end, 0)

	def range_min_query_helper(self, segment_tree, qs, qe, start, end, pos):
		if( qs <= start and qe >= end):
			return segment_tree[pos]

		if(qs > end or qe < start):
			return sys.maxint

		mid = (start+end)//2

		return min(self.range_min_query_helper(segment_tree, qs, qe, start, mid, 2*pos +1),self.range_min_query_helper(segment_tree, qs, qe, mid+1, end, 2*pos +2))



if __name__ == "__main__":
	st_instance = SegmentTree()
	st = st_instance.initialize_segment_tree([1,4,2,3,6,5,6])
	print st_instance.range_min_query(st, 0, 6, 1, 3)

File: SegmentTrees/SegmentTree.java
public class SegmentTree{

	private static int getNextOf2(int n){
		int count = 0;
		while( n != 0){
			n = n >> 1;
			count++;
		}
		return (int)Math.pow(count, 2);
	}

	private static int [] initializeSegmentTree(int [] arr){
		int n = getNextOf2(arr.length);
		int [] segmentTree = new int[2*n-1];


		constructSegmentTree(segmentTree, arr, 0, arr.length-1, 0);
		return segmentTree;
	}

	private static void constructSegmentTree(int [] segmentTree, int [] input, int start, int end, int pos){
		if(start == end){
			segmentTree[pos] = input[start];
			return;
		}

		int mid = (start + end)/2;

		// recurse to the left and the right
		constructSegmentTree(segmentTree, input, start, mid, 2*pos+1);
		constructSegmentTree(segmentTree, input, mid+1, end, 2*pos+2);

		segmentTree[pos] = Math.min(segmentTree[2*pos+1],segmentTree[2*pos+2]);
		return;
	}

	private static int rangeMinQuery(int [] segmentTree, int qStart, int qEnd, int start, int end, int pos){
		if(qStart <= start && qEnd >= end){
			return segmentTree[pos]; // complete overlap
		}

		if(qStart > end || qEnd < start){
			return Integer.MAX_VALUE; // non overlap
		}
		
		int mid = (start + end)/2;
		return Math.min(
			rangeMinQuery(segmentTree, qStart, qEnd, start, mid, 2*pos+1),
			rangeMinQuery(segmentTree, qStart, qEnd, mid+1, end, 2*pos+2)
		);
	}


	public static void main(String [] args){
		int [] arr = new int[]{1,3,5,2,6,7,23,12,9};
		int [] segmentTree = initializeSegmentTree(arr);
		System.out.println(rangeMinQuery(segmentTree, 6, 7, 0, arr.length-1, 0));
		
	}
}